apps/browser/src/key-management/biometrics/background-browser-biometrics.service.ts

```
@@ -1,36 +1,125 @@
-    await this.nativeMessagingBackground().send({ command: "biometricUnlock" });
-    const response = await responsePromise;
-    return response.response === "unlocked";
+  async authenticateWithBiometrics(): Promise<boolean> {
+    try {
+      await this.ensureConnected();
+
+      if (this.nativeMessagingBackground().isConnectedToOutdatedDesktopClient) {
+        const response = await this.nativeMessagingBackground().callCommand({
+          command: BiometricsCommands.Unlock,
+        });
+        return response.response == "unlocked";
+      } else {
+        const response = await this.nativeMessagingBackground().callCommand({
+          command: BiometricsCommands.AuthenticateWithBiometrics,
+        });
+        return response.response;
+      }
+    } catch (e) {
+      return false;
+    }
   }
 
-  async isBiometricUnlockAvailable(): Promise<boolean> {
-    const responsePromise = this.nativeMessagingBackground().getResponse();
-    await this.nativeMessagingBackground().send({ command: "biometricUnlockAvailable" });
-    const response = await responsePromise;
-    return response.response === "available";
+  async getBiometricsStatus(): Promise<BiometricsStatus> {
+    try {
+      await this.ensureConnected();
+
+      if (this.nativeMessagingBackground().isConnectedToOutdatedDesktopClient) {
+        const response = await this.nativeMessagingBackground().callCommand({
+          command: BiometricsCommands.IsAvailable,
+        });
+        const resp =
+          response.response == "available"
+            ? BiometricsStatus.Available
+            : BiometricsStatus.HardwareUnavailable;
+        return resp;
+      } else {
+        const response = await this.nativeMessagingBackground().callCommand({
+          command: BiometricsCommands.GetBiometricsStatus,
+        });
+
+        if (response.response) {
+          return response.response;
+        }
+      }
+      return BiometricsStatus.Available;
```

if we get bad response does this result in "Available" ?

---

I'm honestly not sure why it is written this way; I don't think response.response should ever be empty. Do you think it's better to just throw an error here to make sure we catch these cases?